Mathlib/Combinatorics/Enumerative/Polya.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2025 Zihui Bai. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zihui Bai, Zhengfeng Yang
+-/
+
+module
+
+public import Mathlib.Data.Fintype.Basic
+public import Mathlib.GroupTheory.GroupAction.Basic
+public import Mathlib.GroupTheory.GroupAction.Quotient
+
+/-!
+## Main definitions and results
+
+* `MulAction (Equiv.Perm X) (X → Y)`: action of permutations on colorings by precomposition.
+* `coloringEquiv`: two colorings are equivalent if they lie in the same orbit.
+* `coloringEquiv_equivalence`: this defines an equivalence relation.
+* `orbit_size_eq_index`: the orbit–stabilizer theorem reformulated for colorings.
+
+## Implementation notes
+
+We model a coloring as a function `X → Y`, where `X` is a finite set of objects
+and `Y` is a finite set of colors.
+The permutation group `Equiv.Perm X` acts on colorings by precomposition.
+-/
+
+@[expose] public section
+
+open scoped BigOperators
+open MulAction Finset Equiv
+
+namespace DistinctColorings
+
+variable {X Y : Type*} (c : X → Y)
+
+/--
+A *coloring* is a function `X → Y`.
+The permutation group `Equiv.Perm X` acts on colorings by precomposition:
+`(g • c) x = c (g⁻¹ x)`.
+-/
+instance : MulAction (Equiv.Perm X) (X → Y) where
```

Comment:
The instance creates a diamond in the case `Y=X` once you import `Mathlib.Algebra.Group.Action.Pi`, and you should use [DomMulAct](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/GroupAction/DomAct/Basic.html#DomMulAct) here: once you import `Mathlib.GroupTheory.GroupAction.DomAct.Basic` you can just `inferInstance`.